LoRaEngine/LoraKeysManagerFacade/DeviceGetter.cs

```diff
@@ -153,5 +168,61 @@ public async Task<List<IoTHubDeviceInfo>> GetDeviceList(string devEUI, string ga
 
             return results;
         }
+
+        private async Task<JoinInfo> GetJoinInfoAsync(string devEUI, string gatewayId, ILogger log)
+        {
+            var cacheKeyJoinInfo = string.Concat(devEUI, ":joininfo");
+            var lockKeyJoinInfo = string.Concat(devEUI, ":joinlockjoininfo");
+            JoinInfo joinInfo = null;
+
+            if (await this.cacheStore.LockTakeAsync(lockKeyJoinInfo, gatewayId, TimeSpan.FromMinutes(5)))
+            {
+                try
+                {
+                    joinInfo = this.cacheStore.GetObject<JoinInfo>(cacheKeyJoinInfo);
+                    if (joinInfo == null)
+                    {
+                        joinInfo = new JoinInfo();
+
+                        var device = await this.registryManager.GetDeviceAsync(devEUI);
+                        if (device != null)
+                        {
+                            joinInfo.PrimaryKey = device.Authentication.SymmetricKey.PrimaryKey;
+                            var twin = await this.registryManager.GetTwinAsync(devEUI);
+                            const string GatewayIdProperty = "GatewayID";
+                            if (twin.Properties.Desired.Contains(GatewayIdProperty))
+                            {
+                                joinInfo.DesiredGateway = twin.Properties.Desired[GatewayIdProperty].Value as string;
+                            }
+                        }
+
+                        this.cacheStore.ObjectSet(cacheKeyJoinInfo, joinInfo, TimeSpan.FromMinutes(60));
+                        log?.LogDebug("updated cache with join info '{key}':{gwid}", devEUI, gatewayId);
+                    }
+
+                    if (string.IsNullOrEmpty(joinInfo.PrimaryKey))
+                    {
+                        throw new JoinRefusedException("Not in our network.");
+                    }
+
+                    if (!string.IsNullOrEmpty(joinInfo.DesiredGateway) && gatewayId != joinInfo.DesiredGateway)
```

Comment:
Shouldn't we compare gatewayID ignoring case?

Reply:
yes

Reply:
allright. fixing this.